src/os/bluestore/BlueStore.cc

```diff
@@ -11776,11 +11841,15 @@ int BlueStore::read(
     if (offset == length && offset == 0)
       length = o->onode.size;
 
-    r = _do_read(c, o, offset, length, bl, op_flags);
+    r = _do_read(c, o, offset, length, bl, op_flags, 0,
+          might_need_reformatting ? &span_stat : nullptr);
     if (r == -EIO) {
       logger->inc(l_bluestore_read_eio);
     }
   }
+  if (might_need_reformatting) {
```

Comment:
This should be under same lock as `_do_read()` is executed.
Taking collection lock later is not ok, since some transaction could be injected in-between.

Reply:
Good catch, thanks! This unfortunately results in a need for unique lock for that "reading which could trigger reformatting" instead of shared one.
In other words - scrubbing reads have to be performed under unique lock if we expect following reformatting.

Reply:
Done.